ledger/accountdb.go

```diff
@@ -1026,74 +1080,395 @@ func accountsPutTotals(tx *sql.Tx, totals ledgercore.AccountTotals, catchpointSt
+				gi, ai := 0, 0
+				for _, aidx := range deleted {
+					gi, ai = pad.ExtendedAssetHolding.FindAsset(aidx, gi)
+					if gi == -1 || ai == -1 {
+						return updatedAccounts, fmt.Errorf("failed to find asset group for %d", aidx)
+					}
+					// group data is loaded in accountsLoadOld
+					key := pad.ExtendedAssetHolding.Groups[gi].AssetGroupKey
+					if pad.ExtendedAssetHolding.Delete(gi, ai) {
+						// the only one asset was in the group, delete the group
+						qaed.Exec(key)
+					}
+				}
+			}
+
+			if len(created) > 0 {
+				// sort created, they do not exist in old
+				sort.SliceStable(created, func(i, j int) bool { return created[i] < created[j] })
+				pad.ExtendedAssetHolding.Insert(created, delta.new.Assets)
+			}
+
+			// update DB
+			var result sql.Result
+			for i := 0; i < len(pad.ExtendedAssetHolding.Groups); i++ {
+				if pad.ExtendedAssetHolding.Groups[i].Loaded() {
+					if pad.ExtendedAssetHolding.Groups[i].AssetGroupKey != 0 { // existing entry, update
+						_, err = qaeu.Exec(
+							pad.ExtendedAssetHolding.Groups[i].Encode(),
+							pad.ExtendedAssetHolding.Groups[i].AssetGroupKey)
+						if err != nil {
+							break
+						}
+					} else {
+						// new entry, insert
+						result, err = qaei.Exec(pad.ExtendedAssetHolding.Groups[i].Encode())
+						if err != nil {
+							break
+						}
+						pad.ExtendedAssetHolding.Groups[i].AssetGroupKey, err = result.LastInsertId()
+						if err != nil {
+							break
+						}
+					}
+				}
+			}
+			// reset the cache in old.pad
+			pad.AccountData.Assets = nil
+		}
+	}
+	if err == nil {
+		var rowsAffected int64
+		normBalance := delta.new.NormalizedOnlineBalance(proto)
+		result, err := qabu.Exec(normBalance, protocol.Encode(&pad), delta.old.rowid)
+		if err == nil {
+			// rowid doesn't change on update.
+			updatedAccounts[updateIdx].rowid = delta.old.rowid
+			updatedAccounts[updateIdx].pad = pad
+			rowsAffected, err = result.RowsAffected()
+			if rowsAffected != 1 {
+				err = fmt.Errorf("failed to update accountbase row for account %v, rowid %d", addr, delta.old.rowid)
+			}
+		}
+	}
+
+	return updatedAccounts, err
+}
+
+type accountsNewQueries struct {
+	insertStmt          *sql.Stmt
+	updateStmt          *sql.Stmt
+	deleteByRowIDStmt   *sql.Stmt
+	insertGroupDataStmt *sql.Stmt
+	updateGroupDataStmt *sql.Stmt
+	deleteGroupDataStmt *sql.Stmt
+	queryGroupDataStmt  *sql.Stmt
+}
+
+func makeAccountsNewQueries(db db.Queryable) (q accountsNewQueries, err error) {
+	if q.insertStmt, err = db.Prepare("INSERT INTO accountbase (address, normalizedonlinebalance, data) VALUES (?, ?, ?)"); err != nil {
+		q.close()
 		return
 	}
-	defer deleteByRowIDStmt.Close()
 
-	insertStmt, err = tx.Prepare("INSERT INTO accountbase (address, normalizedonlinebalance, data) VALUES (?, ?, ?)")
-	if err != nil {
+	if q.updateStmt, err = db.Prepare("UPDATE accountbase SET normalizedonlinebalance = ?, data = ? WHERE rowid = ?"); err != nil {
+		q.close()
```

Comment:
instead of repeating the q.close() for each entry, you could do
```
defer func() {
    if err != nil {
        q.close()
    }
}()
```

Reply:
done